### ISSUE TITLE
simple UDRI with examples

### ISSUE THREAD

**Issue by vasslitvinov** [chapel-lang/chapel#9879]
This is a simplification of #9542 .

## Interface

```chpl
record ReduceOp {

  // Create the global op record for reducing values of 'inputType'.
  // Not used when the user provides their own instance.
  // This could be the compiler-generated default constructor,
  // if applicable.
  proc init(type inputType)

  // Create a per-task op given 'parentOp' in the parent task.
  // This could be the compiler-generated copy initializer.
  proc init(parentOp :ReduceOp)

  // Create a per-task accumulation state.
  proc newAccumState() :accumType

  // Accumulate an input value onto the current task's
  // accumulation state 'AS'.
  proc accumulate(ref AS :accumType, input :inputType) :void

  // Combine the parent task's accumulation state 'parentAS'
  // with the child task's accumulation state 'childAS',
  // storing the result back into 'parentAS'.
  // Optional. If not defined, accumulate(parentAS, childAS) is used.
  proc combine(ref parentAS :accumType, childAS :accumType) :void

/* The remaining requirements depend on whether globalAS (the global
   accumulation state) needs to be post-processed into the reduction
   result after all values to be reduced have been processed.

 Case "no generate" : when the final globalAS already IS
      the result of the reduction.

     If so, generate(globalAS) must NOT be defined.
     The requirements below do NOT apply.
     In the case of a reduce intent, the outer variable is used
     for globalAS.

 Case "with generate" : when globalAS needs to be post-processed
      to generate the reduction result.

     If so, generate(globalAS) must be defined.
     The requirements below apply.
*/

  // Generate the reduction result a.k.a output given the global
  // accumulation state 'globalAS', after all values to be reduced
  // have been processed.
  proc generate(globalAS :accumType) :outputType

  // Generate the reduction result in the case of a reduce intent
  // into the outer variable 'outerVar'.
  // Note that semantics of the reduce intent requires
  // that the initial value of the outer variable
  // be reflected in the result of the reduction.
  // Optional when outputType == inputType. If not defined,
  // the following is used:
  //   acumulate(globalAS, output); output = generate(globalAS);
  proc generate(ref output :outputType, globalAS :accumType) :void

}
```

* `inputType` - values being reduced
* `accumType` - per-task accumulation state
* `outputType` - result of the reduction

* "global" - at the outer level - at the same scope as the forall loop or the reduce expression
* "parent", "child" - when a "parent" task creates a "child" sub-task

## Sum reduction

```chpl
record SumReduceOp {
  type inputType;
  proc newAccumState()         { var x: inputType; return x; }
  proc accumulate(ref AS, val) { AS += val; }
}
```

## MaxLoc reduction, no uninit

```chpl
record maxloc {
  type inputType;
  proc newAccumState() return (min(inputType(1)), max(inputType(2)));

  proc accumulate(ref AS, (val,loc)) {
    if AS(1) > val || (AS(1) == val && AS(2) < loc) then
      AS = (val,loc);
  }
}
```

## MaxLoc reduction, with uninit

```chpl
// This is the accumulation state.
record ValIdxRec {
  var value;
  var idx;
  var uninit: bool;

  proc init(type inputType) {
    value  = min(inputType(1));
    idx    = min(inputType(2));
    uninit = true;
  }

  proc init(type inputType, (val,loc)) {
    value  = val;
    idx    = loc;
    uninit = false;
  }

  proc add((val,loc)) {
    if uninit || val > value ||
      (val == value && idx < loc)
    {
      value = val;
      idx = loc;
    }
    uninit = false;
  }
}

record maxloc {
  type inputType;

  proc newAccumState() {
    return new ValIdxRec(inputType);
  }

  proc accumulate(ref AS, (val,loc)) {
    AS.add((val,loc));
  }

  proc combine(ref parentAS, childAS) {
    // Nothing to do if child is uninitialized.
    if !childAS.uninit then
      parentAS.add((childAS.value, childAS.idx));
  }

  proc generate(ref globalAS) {
    // Do this even if it is uninitialized.
    return (globalAS.value, globalAS.idx);
  }
}
```

## SlowSum reduction

```chpl
// This is the accumulation state.
record ArrRec {
  var exps: domain(int);
  var vals: [exps] real;
  proc add(exp: c_int, val: real) {
    // "on this" ?
    if !exps.member(exp) then
      exps += exp;
    vals[exp] += val;
  }
}

// Helper.
proc val2exp(val: real) {
  extern proc frexp(value: real, ref exp: c_int): real;
  var exp: c_int;
  frexp(val, exp);
  return exp;
}

// The reduce op type.
record SlowSumOp {

  proc init(type inputType) { compilerAssert(inputType == real); }

  proc newAccumState()  return new ArrRec();

  proc accumulate(ref AS: ArrRec, val: real) {
    AS.add(val2exp(val), val);
  }

  proc combine(ref parentAS: ArrRec, childAS: ArrRec) {
    for (exp, val) in zip(childAS.exps, childAS.vals) do
      parentAS.add(exp, val);
  }

  proc generate(globalAS: ArrRec) {
    var sum = 0.0;
    for val in globalAS.vals do
      sum += val;
    return sum;
  }
}
```

**Comment by bradcray** [comment]
Here are some initial comments:

* In your definition of the interface, I think the methods would be much clearer if you were to specify the expected formal types of the arguments and return types of the routines (e.g., inputType, accumType, outputType).  This would help with understanding the three versions of accumulate, for example.

* In all these examples, I'd leave initializers out for simplicity if they're identical to the compiler-generated initializers (most of them seem to be).  E.g.,:

```chapel
  proc init(type inputType) {
    this.inputType = inputType;
  }
```

* At first I didn't understand the need for the two definitions of identity(), which makes me worry that users won't either.  Do we expect that the version with this signature:

```chapel
proc identity(outerVar)
```

would ever do anything other than return `outerVar`?  Or, is it responsible for converting the input type into the accumulation type?  Am I correct that slowSum() above would currently drop the initial value of the result var on the floor if it were used as a reduce intent?  How would we write it if we wanted to avoid that?  Bottom line: I find the technique for folding in the initial value to be reasonably unintuitive.

* w.r.t. identity(), my intuition is that it would be more natural to supply a routine that says how to initialize the accumulation type.  Specifically, I'd propose a type method that tells the compiler what the accumulation state type is and a routine to initialize a value of that type instead.  I.e.,:

```chapel
  proc accumType type {
    return accumType;
  }

  proc initAccum(ref AS: accumType) {
    ...
  }
```

I'd also permit `initAccum()` to be omitted in the event that the type's default initialization is sufficient (as it is for the sum and SlowSum cases, I think?).  Otherwise, I worry that if we did a partial reduction of slowSums, we'd create an array of records with associative domain and array fields using the default initializers, then for each element in that array, create a new copy of this record with new associative domain/array fields and then do assignments between the records, resulting in domain assignment and array reallocation—all seeming very expensive.  Using the interface proposed just above, it seems that at the minimum we could do an in-place update to the record if the default initializer is not sufficient, and nothing if it is.

* I'd probably drop childOp() until we have an example that needs it (and if we have such an example, I'd add it to the issue here).  More generally, I think every routine in the proposal should have a reduceOp example that motivates it.

* The three versions of accumulate() seem like overkill to me.  What's the difference between the third version and the combine() routine?

**Comment by bradcray** [comment]
I think there's also a simple typo in your maxloc version that uses a record:

```chapel
proc init(type inputType) {
    value  = max(inputType(1));
```

should be min?

**Comment by vasslitvinov** [comment]
Responding to @bradcray comments:

* I added types to the methods, removed unnecessary initializers, fixed the typo - thank you.

* I added a "User Guide" that navigates the user through the optional/conditional requirements. It still does not explain what those arguments do and what each method returns, however I hope that this will be intuitive (for us, anyway) in most cases.

* The three versions of `accumulate` could be the same method, applicable in the different scenarios as desired, possibly by being generic and allowing different instantiations. If a separate method needs to be written to support combining (for example), then `combine` should be written instead.

* I renamed `identity` to `newAccumState`. More of a mouthful for a simple case, however I feel it is justified.

* Re `identity(outerVar)` - hopefully renaming it to `newAccumState` makes it clearer. Its job is to create an accumulation state with the initial value of the outer variable already incorporated. This will do something other than returning outerVar in the case where accumType differs from outputType.

* BTW I chose against relying on the 2-argument generate() to incorporate outerVar. That's because in some cases generate() will need to ignore the initial value of the `output` argument. For example, when it is used for a partial reduction.

* I prefer not to require the user to provide `accumType` + `initAccum`. This is more work for the user in simple cases and prevents them from relying on type inference to calculate accumType.

> Otherwise, I worry that if we did a partial reduction of slowSums, we'd create an array of records with associative domain and array fields using the default initializers, then for each element in that array, create a new copy of this record with new associative domain/array fields and then do assignments between the records, resulting in domain assignment and array reallocation—all seeming very expensive.

I consider this a shortcoming of our implementation that needs to be addressed. In cases like `var A: [...] ... = someExpr;` , we should initialize each array element to `someExpr` directly, instead of default-initializing it first.

* To justify `newChildOp`, Imagine a min-k reduction. The user would create an instance of a MinK record that keeps the desired value of `k`, either as a field or as an instantiating parameter. The user would pass that instance to the reduce expression or reduce intent. To calculate its childOp, `init(inputType)` is not sufficient as it will not transfer the value of `k`.

**Comment by bradcray** [comment]
* This combination still doesn't sit well with me:

  ```chapel
    // Must be defined when combine(ref parentAS, childAS) is not defined.
    proc accumulate(ref parentAS :accumType, childAS :accumType) :void
  
    // Optional. If not defined, accumulate(ref parentAS, childAS) is used.
    proc combine(ref parentAS :accumType, childAS :accumType) :void
  ```

  Specifically, it seems odd to define an interface by saying "you can define either of these routines, it doesn't really matter which one."  My intuition would be to drop the accumulate version and say:

  ```chapel
    // Required if accumType != inputType.
    // If accumType == inputType, accumulate(parentAS, childAS) will be called
    // when combine() is not defined.
    proc combine(ref parentAS :accumType, childAS :accumType) :void
  ```

  The other option would be to do away with `combine()` altogether and just name all the routines `accumulate()`, but I'm not as fond of this (blurs different things together unnecessarily), and am similarly interested in getting rid of accumulate() for the outerVar case:

* I think having two ways to incorporate the outerVar state seems confusing.  The reason I mentally link it to `generate()` rather than initialization or accumulation is that I view the outerVar as being less about the input type / accumulation type, and more related to the output type and the (potential) need to fold the result of the reduction into the value already stored in the result.  As a specific example, in slowSum(), it seems like a shame to have to insert the value into the associative domain and pull it back out when we could just add it to the result at the end.  Moreover, in a case where the output type truly differs from the input and accumulation types, it seems a shame to have to define how to convert the output type into an accumulation type rather than simply combining two output types (given that the accumulation type has to go in that direction anyway.

  This makes me wonder whether there should be a `fold(result: outputType, outerVar: outputType)` routine that would be used to do this.  And then we could say that if outputType == accumType, we'd call combine() for this purpose; and if outputType == accumType == inputType, we'd use accumulate() for this purpose.  I think the net result of this proposal would be that in the most general case, there would be one newAccumState() routine, one accumulate() routine, one combine() routine, and one fold() routine, but for cases where all the types are identical, all that would be required would be a newAccumState() routine and an accumulate() routine (and if we were to take an approach where the accumulation state's default value was appropriate for a new instance of it, even newAccumState() could be ignored).

  You bring up an interesting point about "perhaps the author of a given reduction would want to be able to ignore the initial value of the output variable."  Defining a fold() routine that simply returned the result and dropped the outerVar on the floor would be a way of doing this.

  This is a pretty different approach than the one you're taking, where the incorporation of outerVar happens at the outset of the computation rather than at the end.  Do you see any downside to switching it?

* it's not clear to me why we'd rather use `newChildOp` than having the author create a copy initializer if they need to pass state from one reduceOp to a child's copy.  It seems like that would take care of the `k` value in your example?

* I still find the case 1 and case 2 descriptions pretty confusing.  What basis should an author use to determine which case they want to build, and where is that documented?  Most of the text I read seemed to say "maybe you're in this case or maybe you're in that one" without helping me determine which case I'd want to write and why (but there's a lot of new text here, so maybe I just missed it...)

* I still think that the set of examples should include a case that motivates every option in the interface to help rationalize its existence.  It appears that currently none of the examples define newChildOp(), outputType, or the two-argument generate() function (and potentially some of the accumulate() / combine() overloads).  I think it'd be easier to defend each choice within the interface with a motivating example that wanted that routine.  Or, put another way, I'm likely to keep advocating for getting rid of unnecessary pieces without an example that motivates its inclusion.

* Belatedly, I'm remembering that we should check to make sure that this interface is sufficient for (and natural for) scans (parallel prefix computations) as well.

**Comment by LouisJenkinsCS** [comment]
I rather like this interface, in particular I can see some very practical use-cases such as gathering results from multiple nodes and tasks into a normally non-thread-safe data structure that relaxes the order of insertions... Could something like this work if, say, the `list` module implemented the same interface (or if any user-data structure does the same)...

```chpl
// Objective: Go from distributed array -> distributed list (no synchronization needed)
var data : list(int);
var A : [someDistributedDomain] int;
forall a in A with (reduce data) {
   data.push_back(a);
}
```

The above would be extremely cool if the data structure can handle this by handling reduction across multiple nodes and tasks... In fact, I believe it shouldn't be restricted to `record` types, but `class` types as well, the types themselves need to just implement the necessary reduction methods. Would the above be possible? I can see it helping a lot for this kind of data-parallelism and non-irregular applications.

**Comment by LouisJenkinsCS** [comment]
I believe another really good example, would be on that @mppf had mentioned in issue #8339...

```chpl
var D:domain(int);
var Count:[D] atomic int;
forall key in input with (reduce Count) {
  Count[key] += 1;
}
```

Of course, in the above we really shouldn't require the user to append the key to the domain each time, it should create it on demand... perhaps for associative arrays they should just have a `insertIfPresent(key, val)` method. That would solve the issue.

**Comment by vasslitvinov** [comment]
I just updated the main text of this issue. Key changes:

* I switched to 2-arg generate() to incorporate the initial value of the outer variable. This simplified the interface. The outer variable's value is now incorporated at the end, rather than at the beginning. I am not aware of any issues this may cause.

* I removed the "User Guide", as it does not seem needed any more. Instead I added comments to the interface itself.

* I kept the name distinction "combine" vs. "accumulate". Dropping combine() in favor of accumulate() is still allowed, as long as accumulate() is applicable to the arguments.

* I still need to check if this works for scans.

**Comment by vasslitvinov** [comment]
For scans - confirming that this interface is sufficient. We can rewrite `chpl__scanIterator()` in ChapelReduce.chpl as follows:

```chpl
  iter chpl__scanIterator(op, data) {
    var AS = op. newAccumState();
    for e in data {
      op.accumulate(AS, e);
// "no generate"
      yield AS;
// "with generate"
      yield op.generate(AS);
    }
    // perhaps op.deinit();
  }
```

**Comment by bradcray** [comment]
> For scans

I should've clarified:  For scans written properly (using the parallel algorithm), not the lame serial algorithm we have today.  See slides 25+ in [this lecture](https://courses.cs.washington.edu/courses/csep524/13wi/Lecture5.pdf) though it only covers the 1D case and we'll want to support the multi-dimensional (partial and full) cases as well.

**Comment by vasslitvinov** [comment]
For parallel scans: in addition to newAccumState + accumulate + combine + generate, there is a lot of passing around of the accumulation state.

That may deserve explicit support for performance is non-trivial cases, such as ArrRec in the SlowSum example above. The default copy operation should be sufficient in the simpler cases like numeric operations. Serialization (a recent feature currently for internal use) may handle some of the more interesting cases. When/if all this fails, we can consider expanding the ReduceOp interface.

**Comment by mppf** [comment]
In the summary in the PR message it says

>   proc newAccumState() :accumType

but this is the first mention of accumType. Is it missing a `type accumType` field?

Re. scans, I think it'd be instructive to write a sketch/psuedocode of the 1-D array parallel scan case.
I can think of 2 versions of this pseudocode:
 1. Tasks compute the total of the group of elements they own & then the running totals among the groups are computed serially & then tasks add the running totals to the group elements they own
 2. As above but the running totals among the groups are computed in parallel a-la a reduction

Actually it seems to me that the problem with parallel scans is likely to be the need to iterate over the data multiple times (which AFAIK our parallel iterators do not currently allow - but maybe they are closer to allowing it than I think) rather than the reduction interface. Even still, I think it's absolutely worth our time to identify what the gaps are.

It occurs to me that I've written this parallel prefix sum before (with MPI), so it might be easy/fun for me to try writing these sketches:

``` chapel
// sketch of scan over array
// working with myReduceOp which has inputType and accumType among other things
// For simplicity I write inputType/accumType and not e.g. myReduceOp.inputType
var A:[1..n] inputType;
var nTasks = getNumTasks(A.domain); // presumably based on min( # cores, n )
var PerTaskTotals:[1..nTasks] accumType; // distributed ?
var barrier1 = new NTaskBarrier(nTasks);
var barrier2 = new NTaskBarrier(nTasks);
coforall task in 1..nTasks {
  // Gather up the total for this task and store it in PerTaskTotals
  var thisTaskTotal:  myReduceOp.newAccumState();
  for elt in iterateTaskChunk(task, nTasks, A) {
    myReduceOp.accumulate(thisTaskTotal, elt);
  }
  PerTaskTotals[task] = thisTaskTotal;
  // Wait for all tasks to have computed their PerTaskTotal
  barrier1.barrier();
  // Now 1 task will serially compute the running totals for the groups
  // After this section, PerTaskTotals[t+1] - PerTaskTotals[t]
  // will store the total from task t. and PerTaskTotals[1] will be "0".
  var allTaskTotal:  myReduceOp.newAccumState();
  if task == 1 {
    var allTaskTotal:  myReduceOp.newAccumState();
    for t in 1..nTasks {
      var thisTaskTotal = PerTaskTotals[t];
      PerTaskTotals[t] = allTaskTotal;
      myReduceOp.combine(allTaskTotal, thisTaskTotal);
    }
  }
  // All tasks wait for the running totals to be computed
  barrier2.barrier();
  // Now each task computes the running total for its elements
  var thisTaskRunningTotal = PerTaskTotals[task];
  for elt in iterateTaskChunk(task, nTasks, A) {
    var originalElement = elt;
    elt = thisTaskRunningTotal;
    MyReduceOp.combine(thisTaskRunningTotal, originalElement);
  }
}
```

Now, I could've written that with two different coforall loops instead of the barriers:
```
coforall task in 1..nTasks {
  // compute each task PerTaskTotals contribution
}
// convert PerTaskTotals into running totals in 1 task
coforall task in 1..nTasks {
  // add PerTaskTotals[task] to each task PerTaskTotals contribution & do running totals
}
```

Either way, I don't think computing the running totals among the group in parallel (sketch 2 above) really changes the picture for the reduction interface. One approach there would be to use a recursive scan to compute the running totals...

What are the real issues with scans, then?
 * Like someIterableExpr.sorted(), they require iterating over someIterableExpr twice
 * What iterator semantics do we provide for parallel iterators that can be run multiple times? Is this confusing to iterator authors? Is it just a matter of fixing up the implementation?
 * If they're to be added in to user forall loops (say with "scan forall intent), we'd presumably have to convert that forall loop into 2 forall loops: the first is the 1st section above & includes the user's forall loop body to generate the element data; the 2nd is a new loop that matches the 2nd section above. But this 2nd loop should presumably match the locality distribution of the 1st forall loop so it might not be an entirely new loop. How can we do that with reasonable language semantics?
 * Even for the standalone scan over an array case, we'll want these loops to match the array distribution, so PerTaskTotals might be distributed or some more complex per-locale then per-task hierarchical data structure. Having a more complex per-locale then per-task hierarchical data structure might also help with creating a scan that combines the running totals among the groups in parallel.

**Comment by vasslitvinov** [comment]
> Is it missing a type accumType field?

`inputType` / `accumType` / `outputType` are not part of the required interface. They are used in the interface to clarify the intention of each method.

**Comment by vasslitvinov** [comment]
> In other words accumulate and generate will have generic arguments instead of using accumType?

It is up to the ReduceOp author. Here are some scenarios:

* Yes, generic arguments. For example: `proc accumulate(AS, input) { AS += input; }`
* No, concrete arguments. In the "SlowSum reduction" example above: `proc accumulate(ref AS: ArrRec, val: real)`
* Have an `inputType` type field in the ReduceOp and require that the arugments be of that type: `proc accumulate(AS: inputType, input: inputType) { AS += input; }`

In other words, the interface does not require specific types on the method declarations. (Or specific intents, for that matter.) Instead it requires that the corresponding calls be resolvable. For example, given `op:ReduceOp`, `AS:accumType`, `input:inputType`, when those are created for a reduction with the given `inputType`, the call `op.accumulate(AS, input)` must be resolvable.

**Comment by vasslitvinov** [comment]
Looking at the Count example by @LouisJenkinsCS [above](https://github.com/chapel-lang/chapel/issues/9879#issuecomment-398864715), I can see two ways to think about it.

#### Each task has its own Count array.

This is the case that the interface is designed for. In this case, Count's elements can be integers. They do not need to be atomic. The code looks like this:

```chpl
var D: domain(int);
var Count: [D] int;
forall key in input with (CountOp reduce Count) {
  Count reduce= input;
}
```

Or, equivalently:

```chpl
var Count = CountOp reduce input;
```

The reduce op is:

```chpl
record CountOp {
  type inputType;
  // see newAccumState()
  var D: domain(inputType);
  // use default initializers

  proc newAccumState() {
    // We could declare 'D' here instead of in CountOp.
    // In that case, however, we couldn't add indices to it.
    // Also I am not sure we can return an array
    // over a local domain.
    //var D: domain(inputType);
    var LocalCount: [D] int;
    return LocalCount;
  }

  proc accumulate(LocalCount, input) {
    D.add(input); // Do we need to check D.member(input) ?
    D[input] += 1;
  }

  proc combine(MyCount, ChildCount) {
    ... add each element in ChildCount to MyCount ...
  }

  // no generate() is necessary
}
```

#### All tasks on a given locale share a Count array.

The per-task operations as as above. As to the per-locale Count array, we want it:

* To have atomic ints as elements, so they can be updated concurrently as Louis suggests.

* To have a parSafe domain, so that indices can be added concurrently.

The key issue here is that the interface does not support the distinction between per-locale and per-task accumulation state and the corresponding accumulate/combine operations.

We could see whether this example can be written using task-private variables. However, currently there is no support for distinguishing between per-locale and per-task variables. So it may be challenging to do it that way, too.

#### Distributed final result.

If the array where we want to have the final counts is distributed, the current proposal will lead to locking it every time a combine() will be invoked from one of the locales. This may not be a perfect fit because we would like to combine per-locale data concurrently.

**Comment by LouisJenkinsCS** [comment]
I'd say that `CountOp` should be offered as a utility reduction operation, as it is bound to be a very commonly desired reduction, and that it gets named something more descriptive like `AssociativeCounter`. I believe all modules should provide their own reductions that can be accessed in a similar fashion.

**Comment by bradcray** [comment]
> I believe all modules should provide their own reductions that can be accessed in a similar fashion.

I'm having trouble understanding this.  What reductions would you expect the FileSystem, DateTime, or Reflection modules to support?

**Comment by LouisJenkinsCS** [comment]
Not literally all modules, just the one's where they make sense, such as `List`, `ChapelRange`, `ChapelArray`, etc. Speaking of which, would syntax like this still work? 

```chpl
var arr : [1..10] int = 1..10;
var maxVal = max reduce arr;
```

In the new reduction? Would hate to have to create a ReduceOp implementation for something so trivial.

**Comment by vasslitvinov** [comment]
Reopening it so we can finalize and implement.
